forecast/clusterer.py

```diff
@@ -236,174 +265,175 @@ def _build_neighbors(centers, timestamps, n_neighbors):
+            if last_cluster_element or still_belongs:
+                #reason = ('L' if last_cluster_element else '') + ('B' if still_belongs else '')
+                #print(f'Template stayed in cluster {old_cluster} because ({reason}): {self._dbgname[template]}')
+                return old_assignment
+
+            # Otherwise, eliminate the template from its old cluster.
+            self._modify_cluster(False, old_assignment, template, start_time, end_time)
+            #print(f'Template eliminated from cluster {old_cluster}: {self._dbgname[template]}')
+
+        new_assignment = None
+        # try to find a cluster membership for the template
+        if neighbors == None:
+            for cluster in self.centers.keys():
+                if self._similarity(
+                    self._query_df(self._df, template, timestamps).values,
+                    self._query_series(self.centers[cluster], timestamps).values
+                ) > self.rho:
+                    new_assignment = cluster
+                    break
+        else:
+            data = self._query_df(self._df, template, timestamps)[
+                'count'].values.reshape(1, -1)
+            data = sklearn.preprocessing.normalize(data)
+            neighbor = neighbors.kneighbors(data, return_distance=False)[0][0]
+            clusters = sorted(self.centers.keys())
+            if self._similarity(data, self.centers[clusters[neighbor]].values) > self.rho:
+                new_assignment = clusters[neighbor]
+
+        # If this template found a cluster to join, then make the assignment and continue.
+        if new_assignment is not None:
+            #description = 'joined' if old_assignment is None else 'reassigned to'
+            #print(f'Template {description} cluster {new_cluster}: {self._dbgname[template]}')
+            self._modify_cluster(True, new_assignment, template, start_time, end_time)
+            return new_assignment
+
+        # Otherwise, this template needs a new cluster. Make a new cluster.
+        new_assignment = self.next_cluster
+        self.next_cluster += 1
+
+        self.centers[new_assignment] = self._query_df_range(
+            self._df, template, start_time, end_time)
+        assert self.centers[new_assignment].index.name == 'log_time_s'
+        assert self.centers[new_assignment].columns.values == ['count']
+        if self.centers[new_assignment].shape[0] == 0:
+            print(
+                f'WARNING: cluster {new_assignment} has no items. Does the following query appear within the lookback window: {self._dbgname[template]}')
+
+        self.cluster_sizes[new_assignment] = 1
+        self.cluster_totals[new_assignment] = 0
+        print(
+            f'Created cluster {new_assignment} based on template: {self._dbgname[template]}')
+        return new_assignment
+
+    def _cluster_online(self):
+        # mapping of cluster id to df representing center of cluster
+        self.centers: Dict[int, pd.DataFrame] = {}
```

Comment:
I think I recall you mentioning this, but I forget exactly why -- what was the main advantage of working mainly with references to a class-scoped `self.centers`, etc., vs working with a locally scoped object and only modifying it later?

Arguments for the latter are mostly around:
- Explicitly identifying all dependencies in e.g., helper methods for a class.
- You can grep quickly for all places where the variable is set; there will be much fewer places.
- Your updates are only done at conceptual boundaries so you are less likely to have inconsistent state.
- In general, a light form of the principle of least privilege.

Reply:
I factored out some of `_cluster_online` originally because the amount of nesting made it difficult to read and understand the code. The helper methods that are being called in `_cluster_online` all end up modifying `centers`, `cluster_sizes`, and `cluster_totals`. I now see why it might be questionable to always modify class-scoped members, but I did not want to have helpers with 10 arguments so we pass in local objects, since this defeats my original purpose for refactoring. But I am definitely open to changing it back.

Reply:
Hm. Could we have a 

```python
@dataclass 
class ClusterObject:
    centers: (the type)
    cluster_sizes: (the type)
```

to wrap bundles of arguments together? Kind of an in-between.

I prefer arguments to be explicit because it limits the amount of code you have to read, but I agree that anything over... 6?... arguments is probably overkill for this kind of code.

Reply:
This is still unresolved; but I am OK with leaving it as it is for now.